### ISSUE TITLE
Multiple LinkAdrRequests after join 

### ISSUE THREAD

**Issue by XKottelaarAlflex** [chirpstack/chirpstack#537]
Hi,

We are having some connectivity issues with our nodes where we are missing uplinks from the node. After some research we found that when the node joins the network it receives multiple LinkAdrRequests in under a minute.

The scenario can be described like the following:
1. Join request
2. Join accept
3. Uplink, application payload, f_cnt: 1
4. Downlink, Link ADR request dr: 5, tx_power: 1
5. Uplink, MAC data only, Link ADR answer, f_cnt: 2
6. Downlink, Link ADR request dr: 5, tx_power: 4
7. Uplink, MAC data only, Link ADR answer, f_cnt: 3
8. Downlink, Link ADR request dr: 5, tx_power: 7
9. No further data until next periodic uplink.

Note that the time it takes from the first uplink (3.) to the last uplink (7.) takes place in about 16 seconds. To my current understanding tx_power: 7 indicates that our node is requested to output 14dBm below our maximum capable transmit power. For our node this would mean that our tx power will be lowered from +14dBm to 0dBm (if my calculations are correct).

Because of how aggressive the transmit power of our node is lowered I went looking into the default ADR algorithm located in chirpstack/src/adr/default.rs. I found that almost exactly matches the algorithm as described in https://www.thethingsnetwork.org/forum/uploads/default/original/2X/7/7480e044aa93a54a910dab8ef0adfb5f515d14a1.pdf. Although this is an old preliminary document but as of current I couldn't find a newer version of it on the Semtech website.

I mentioned that it almost matched the reference algorithm because the Chirpstacks implementation seems to be not taking into account that the algorithm requires at least 20 uplinks as an input with frame counters from Fn to Fn-19 however in the scenario described earlier this doesn't seem to be the case. The ADR algorithm is executed while the frame counter hasn't exceeded 20 frames.

From reviewing the default.rs the 20 uplinks seem to be only taken into account when increasing the spreading factor (decreasing the data rate) to prevent oscitation. However the document describes that the algorithm requires 20 uplink frames. The document also describes a strategy on when to send the LinkAdrRequest, one of which is to only send an ADR request when the node sets the ADR_ACK_REQ bit and the other describes to send the request every Nth frame. 

I would expect that after a Join we would only receive an ADR command to update the chanel mask but the next ADR message to actually change the data rate or tx power of the end node after the uplink frame counter has exceeded 20 (at least according to the documentation I have available). The next ADR change I would then expect to be 20 uplinks after (fCnt >= 40) the first etc. (or when ADR_ACK_REQ is set). At least when Chiprstack has enough data to safely increase the datarate and lower the tx power.

Also after a join I would expect the history to be cleared but I'm currently unsure if this is true. 

Is it possible to change the ADR algorithm to use apply the 20 frame history after a join? Or am I missing something else?

We are running Chirpstack 4.9.0 on the network server.
A port of https://github.com/Lora-net/SWL2001/releases/tag/v4.3.0 on the node.

I've atached a log file that contains more information releated to this issue.

[log adr on join.json](https://github.com/user-attachments/files/17365444/log.adr.on.join.json)

**Comment by brocaar** [comment]
> At least when Chiprstack has enough data to safely increase the datarate and lower the tx power.

Please note that the tables contains the best SNR of each uplink (in case reported by multiple gateways). Then the best SNR of the last 20 items is used to calculate the number of data-rate increases and / or tx-power decreases. ChirpStack resets the table on every ADR change, as an ADR change invalidates previously collected values.

Therefore I believe it doesn't matter to wait until the table is full or not before starting the algorithm, because it will only get more optimistic (increasing the data-rate / lowering the tx-power even more), not less.

---

The `AdrACKReq` bit (now) serves a different purpose. If ADR is enabled at the end-device, and it doesn't receive a downlink for `ADR_ACK_LIMIT` uplinks, it will set this flag to request a downlink from the network-server, even if it is empty to confirm that the device is still connected.

**Comment by XKottelaarAlflex** [comment]
> Therefore I believe it doesn't matter to wait until the table is full or not before starting the algorithm.

I understand that if the ADR algorithm works on the best SNR it shouldn't really matter if you wait until the table is full or on the next uplink that happens to have a better SNR. However in our case an ADR request triggers an instant reply when received by the node. The chirpstack then uses this ADR reply to again determine a new DR and tx power limit. This proces repeats until there is not enough margin left to lower the datarate or tx power. So instead of having a table with data that is filled with due to the behavior of the node (it's send interval or events), it's being filled with ADR replies that are received within a minute after joining. The ADR algorithm is now working on data that would otherwise not be available if the node didn't answer instantly.

> because it will only get more optimistic (increasing the data-rate / lowering the tx-power even more), not less.

Sorry I don't really understand, how having more samples over a prolonged time at a static datarate and tx power would lead to a more optimistic result? Or do you mean that over time, even using a full table the node would eventually get an ADR request to the same datarate and tx power as it did after joining?

> The AdrACKReq bit (now) serves a different purpose. If ADR is enabled at the end-device, and it doesn't receive a downlink for ADR_ACK_LIMIT uplinks, it will set this flag to request a downlink from the network-server, even if it is empty to confirm that the device is still connected.

Totally agree, I think the document doesn't intent to replace that behavior but rather adding to it. So in addition to the (possible empty) downlink also use it as a moment to run the ADR algorithm and send a ADR request.

**Comment by brocaar** [comment]
> Or do you mean that over time, even using a full table the node would eventually get an ADR request to the same datarate and tx power as it did after joining?

Correct. The ADR algorithm uses the best SNR from the table, thus more items would never result in a lower data-rate. 

> I think the document doesn't intent to replace that behavior but rather adding to it.

The `ADR_ACK_LIMT` is 64. Which means this `ADRACKReq` bit is only set in case there have been 64 uplinks without any (data) downlink. Thus if there are regular downlinks, the `ADRACKReq` bit might never been set during the lifetime of the device. Thus this is not a good trigger for ADR mac-commands.

> However in our case an ADR request triggers an instant reply when received by the node.

Please note that this is end-device application specific. Not all nodes will immediately respond to mac-commands. Some devices will postpone the confirmation until the next uplink.

I'm going to close this issue because I do not see a reason to change the implementation. I do however understand that the default implementation does not suit all implementation. Please note that it is possible to implement your own ADR algorithm (in JS). Please see the ChirpStack configuration options. Here is an example skeleton:

https://github.com/chirpstack/chirpstack/blob/master/examples/adr_plugins/plugin_skeleton.js
